### PR TITLE
feat(eslint-plugin): [restrict-template-expressions] add option to allow RegExp

### LINKED ISSUE #3671: [restrict-template-expressions] add option to allow regular expressions
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/restrict-template-expressions": ["error"]
  }
}
```

```TS
function log(value: RegExp) {
  console.log(`${value}`);
}
```

**Expected Result**

Since `RegExp` defines `.toString()`, I would expect that there would be options to allow printing them similar to `allowBoolean`.

**Actual Result**

No way to make this allowed without disabling the rule.

**Additional Info**

Context: hoping to enable this internally for #3278.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.28.5` |
| `@typescript-eslint/parser`        | `4.28.5` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.27.0` |
| `node`                             | `16.6.0` |

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #3671

### CODE REVIEW

File: packages/eslint-plugin/src/rules/restrict-template-expressions.ts [hunk 1]
```diff
@@ -76,6 +78,13 @@ export default util.createRule<Options, MessageId>({
         return true;
       }
 
+      if (
+        options.allowRegExp &&
+        util.getTypeName(typeChecker, type) === 'RegExp'
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/restrict-template-expressions.ts:83]: Why not `util.isTypeFlagSet`?

Comment (a-tarasyuk) [reply packages/eslint-plugin/src/rules/restrict-template-expressions.ts:83]: https://github.com/microsoft/TypeScript/blob/main/src/compiler/types.ts#L5048 doesn't contain flags for `RegExp`.

**Review by Caglar1881 — CHANGES_REQUESTED** [review]
👍

**Review by Caglar1881 — COMMENTED** [review]
Başarı

**Review by Caglar1881 — CHANGES_REQUESTED** [review]
Sen

**Review by bradzacher — APPROVED** [review]
thanks for this!

**Pull request merged**